### PR TITLE
Automatic Nightly Patch Releases

### LINKED ISSUE #46231: Nightly Patch Releases
see discussion

- New githib action: Release-1b-auto-patch
    - [ ] runs nightly on a cron
    - [ ] optional input: major release (for on-demand patch)
    - [ ] find last commit with green CI for last 2 releases (eg 49 and 50)
    - [ ] make sure that commit hasn't been released
    - [ ] find next patch version based on tags
    - [ ] trigger tag-for-release job with a new flag: `auto`
- `auto` flag:
    - [ ] gets passed from tag job to test job
    - [ ] test job with `auto` flag triggers release job
        - [ ] skips release notes
        - [ ] skips version-info update
        - [ ] skips milestone management
        - [ ] skips docs updates
        - [ ] s3 and docker do not tag patch releases as `latest`
        - [ ] sends flag to ee-extra
            - [ ] ee-extra auto approves/merges `auto` flag PRs with green CI

### PR DESCRIPTION
**Pull request by iethree**
Closes https://github.com/metabase/metabase/issues/46231

### Description

Screen Shot 2024-07-29 at 4 54 09 PM

**1. New workflow: Auto Patch ([sample run](https://github.com/iethree/metabase/actions/runs/10152755714))**
- runs nightly on weekdays, or can be triggered to run manually for a single version
- scheduled runs attempt to release from 2 recent release branches
- checks git tags to find the last released version on that branch and increments by one patch
- checks the commits on the branch to find the last one with green CI
- if the last green commit hasn't been released yet, it triggers the existing build job with a new `auto` flag

**2. tag-for-release workflow update** ([sample run](https://github.com/iethree/metabase/actions/runs/10152759657))
- passes through the `auto` flag 

**3. pre-release testing workflow update** ([sample run](https://github.com/iethree/metabase/actions/runs/10152782187))
  - if the `auto` flag is present, it automatically triggers the publish job upon successful completion 

**4. Publish workflow update** ([sample run](https://github.com/iethree/metabase/actions/runs/10152995787))
  - receives the new `auto` tag, and for auto-releases:
	  - does not generate release notes
	  - does not update version-info.json
	  - does not push latest tagged docker image
	  - does not trigger docs updates
	  - does not close/open any milestones
	  - has a slightly different end-of-release slack message ([example](https://metaboat.slack.com/archives/C01BWAZJE0N/p1722293519946239?thread_ts=1722292848.599809&cid=C01BWAZJE0N))

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: .github/workflows/release-patch.yml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+on:
+  workflow_dispatch:
+    inputs:
+      version:
+        description: 'Major Metabase version (e.g. 45, 52, 68)'
+        type: number
+        required: true
+  schedule:
+    - cron: '0 7 * * 1-5' # every weekday at 8am/9am Central European Time
```
Comment (iethree) [inline .github/workflows/release-patch.yml:12]: @perivamsi or @paoliniluis any preferences for when the nightly build should happen?

Comment (perivamsi) [reply .github/workflows/release-patch.yml:12]: 8 pm ET

File: .github/workflows/pre-release.yml [hunk 2]
```diff
@@ -285,6 +289,30 @@ jobs:
         run: while ! curl -s 'http://localhost:3000/api/health' | grep '{"status":"ok"}'; do sleep 1; done
         timeout-minutes: 3
 
+  trigger-auto-publish:
+    if: ${{ inputs.auto }}
```
Comment (uladzimirdev) [inline .github/workflows/pre-release.yml:293]: iirc we need to compare with strings like `inputs.auto != 'false'`, but maybe `default` makes it differently

https://stackoverflow.com/a/76294014

File: .github/workflows/pre-release.yml [hunk 3]
```diff
@@ -285,6 +289,30 @@ jobs:
...
+    permissions:
+      actions: write
+    runs-on: ubuntu-latest
+    timeout-minutes: 5
+    steps:
+      - name: trigger auto publish
+        uses: actions/github-script@v7
+        with:
+          script: | # js
+            await github.rest.actions.createWorkflowDispatch({
```
Comment (uladzimirdev) [inline .github/workflows/pre-release.yml:304]: nit: we can probably use `gh`, but I'm fine with rest api if top level await is supported

File: .github/workflows/release-patch.yml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+              });
+
+              const goodCommit = await getLatestGreenCommit({
+                github,
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+                branch: `release-x.${majorVersion}.x`,
+              });
+
+              const hasBeenReleased = await  hasCommitBeenReleased({
```
Comment (uladzimirdev) [inline .github/workflows/release-patch.yml:52]:
```suggestion
              const hasBeenReleased = await hasCommitBeenReleased({
```

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -31,6 +31,10 @@ on:
         description: 'Skip publishing release notes'
         type: boolean
         default: false
+      auto:
+        description: 'auto-patch release DO NOT SET MANUALLY'
```
Comment (uladzimirdev) [inline .github/workflows/release.yml:35]:
```suggestion
        description: 'auto-patch release DO NOT SET MANUALLY OR WILL BE FIRED!'
```

File: release/src/github.ts [hunk 6]
```diff
@@ -192,3 +192,94 @@ export async function getIssueWithCache ({
 
   return issue?.data as Issue | null;
 }
+
+function checksPassed(checks: GithubCheck[]) {
+  const MIN_TOTAL_CHECKS = 95;
```
Comment (uladzimirdev) [inline release/src/github.ts:197]: but why?

File: release/src/slack.ts [hunk 7]
```diff
@@ -22,7 +22,7 @@ export function mentionUserByGithubLogin(githubLogin?: string | null) {
   if (githubLogin && githubLogin in githubSlackMap) {
     return `<@${githubSlackMap[githubLogin]}>`;
   }
-  return '@unassigned';
+  return `${githubLogin ?? 'unassigned'}`;
```
Comment (uladzimirdev) [inline release/src/slack.ts:25]:
```suggestion
  return `${githubLogin ?? '@unassigned'}`;
```

?

**Review by uladzimirdev — APPROVED** [review]
🤞 

comments inline

File: .github/workflows/release.yml (hunk 5 above)
Comment (iethree) [reply .github/workflows/release.yml:35]: 🤣

File: release/src/github.ts (hunk 6 above)
Comment (iethree) [reply release/src/github.ts:197]: Magic!

v49 has this many checks. I'll make this clearer.

File: .github/workflows/pre-release.yml (hunk 2 above)
Comment (iethree) [reply .github/workflows/pre-release.yml:293]: the types on these github action params are 🥴 

I ran another test on this workflow, and it's working as intended.

[run without auto flag](https://github.com/iethree/metabase/actions/runs/10199803482):  skips auto-publish
[run with auto flag](https://github.com/iethree/metabase/actions/runs/10152952272): triggers auto-publish

**Pull request merged**